students/annguan/lesson04/mailroom2.py

```diff
@@ -0,0 +1,148 @@
+#!/usr/bin/env python3
+
+### Lesson 4 Mailroom, Part 2, uses dict
+
+from textwrap import dedent
+import os
+import sys
+import math
+
+# a data structure holds a list of donors and a history of the amounts they have donated.
+def get_donor_db():
+    return {'Abraham Lincoln': ("Abraham Lincoln", [145674.32, 23465]),
+'Barack Obama': ("Barack Obama", [3456324.11, 3495, 323]), 
+'Charlie Brown':("Charlie Brown",[453.67]),
+'Doctor Who': ("Doctor Who", [5600, 42]),
+'Eve WallE': ("Eve WallE",[2.22])}
+
+# Write thank you letters to disk as a text file
+def write_letters_disk():
+    for donor in donor_db.values():
+        letter = comp_letter(donor)
+        filename = donor[0].replace(" ","_")+".txt"
+        open(filename,'w').write(letter)
+
+# prompt user to choose from a menu of 3 actions: "Send a Thank you", "Create a Report", or "quit"
+def menu_selection():
+    action = input(dedent('''
+        Choose an action:
+        '1' - Send letters to all
+        '2' - Send a Thank You
+        '3' - Create a Report
+        '4' - Quit
+        >'''))
+    return action.strip()
+
+# Thank You Letter Composition
+def comp_letter(donor):
+    return dedent('''
+        Dear {0:s}
+
+        Thank you for your donation of ${1:.2f}.
+
+        See you soon.
+
+        '''.format(donor[0], donor[1][-1]))
+
+#Validate donation inputs by user
+def check_donation(donation):
+    donmat = float(donmat)
+    if math.isnan(donmat) or math.isinf(donmat):
+        raise ValueError
+    if donmat < 0:
+        raise ValueError ("negative donation")
+    return donmat
+
+
+# Sending a Thank You
+
+def add_donor(name):
+    name = name.strip()
+    donor = (name,[])
+    donor_db[name.lower()] = donor
+    return donor
+
+def find_donor(name):
+    key = name.strip().lower()
+    return donor_db.get(key)
+
+def list_donors():
+    l1 = ["Donor List:"]
+    for donor in donor_db.values():
+        l1.append(donor[0])
+    return "\n".join(l1)
```

Comment:
this can really be `return "\n".join(["Donor List:"] + list(donor_db))`

Reply:
Good idea! Thanks Natasha.